source/custom/nirfmxinstr_restricted_service.custom.cpp

```diff
@@ -15,4 +22,191 @@ ::grpc::Status NiRFmxInstrRestrictedService::ConvertApiErrorStatusForNiRFmxInstr
+//---------------------------------------------------------------------
+::grpc::Status NiRFmxInstrRestrictedService::GetInitiaitedSnapshotStrings(
+  ::grpc::ServerContext* context,
+  const GetInitiaitedSnapshotStringsRequest* request,
+  GetInitiaitedSnapshotStringsResponse* response)
+{
+  if (context->IsCancelled()) {
+    return ::grpc::Status::CANCELLED;
+  }
+  try {
+    auto instrument_grpc_session = request->instrument();
+    niRFmxInstrHandle instrument = session_repository_->access_session(instrument_grpc_session.id(), instrument_grpc_session.name());
+    uInt64 snapshot_info_cache_index {};
+    int32 personality_id_array_actual_size {};
+    int32 signal_names_actual_size {};
+    int32 result_names_actual_size {};
+    int32 snapshot_identifiers_actual_size {};
+    int32 snapshot_timestamp_array_actual_size {};
+    auto status = library_->GetInitiaitedSnapshotStrings(
+      instrument,
+      &snapshot_info_cache_index,
+      nullptr, // personalityIDArray
+      0,       // personalityIDArraySize
+      &personality_id_array_actual_size,
+      nullptr, // signalNames
+      0,       // signalNamesSize
+      &signal_names_actual_size,
+      nullptr, // resultNames
+      0,       // resultNamesSize
+      &result_names_actual_size,
+      nullptr, // snapshotIdentifiers
+      0,       // snapshotIdentifiersSize
+      &snapshot_identifiers_actual_size,
+      nullptr, // snapshotTimestampArray
+      0,       // snapshotTimestampArraySize
+      &snapshot_timestamp_array_actual_size);
+    if (!status_ok(status)) {
+      return ConvertApiErrorStatusForNiRFmxInstrHandle(context, status, instrument);
+    }
+    response->mutable_personality_id_array()->Resize(personality_id_array_actual_size, 0);
+    int32* personality_id_array = reinterpret_cast<int32*>(response->mutable_personality_id_array()->mutable_data());
+    std::string signal_names;
+    if (signal_names_actual_size > 0) {
+        signal_names.resize(signal_names_actual_size - 1);
+    }
+    std::string result_names;
+    if (result_names_actual_size > 0) {
+        result_names.resize(result_names_actual_size - 1);
+    }
+    std::string snapshot_identifiers;
+    if (snapshot_identifiers_actual_size > 0) {
+        snapshot_identifiers.resize(snapshot_identifiers_actual_size - 1);
+    }
+    response->mutable_snapshot_timestamp_array()->Resize(snapshot_timestamp_array_actual_size, 0);
+    uInt64* snapshot_timestamp_array = reinterpret_cast<uInt64*>(response->mutable_snapshot_timestamp_array()->mutable_data());
+    auto personality_id_array_size = personality_id_array_actual_size;
+    auto result_names_size = result_names_actual_size;
+    auto signal_names_size = signal_names_actual_size;
+    auto snapshot_identifiers_size = snapshot_identifiers_actual_size;
+    auto snapshot_timestamp_array_size = snapshot_timestamp_array_actual_size;
+    status = library_->GetSnapshotInfoFromCache(
+      instrument,
+      snapshot_info_cache_index,
```

Comment:
Calling this method with an index of 0 will return an internal error.  If the above call returns an index of 0, we should skip trying to retrieve the snapshot because there was nothing to cache.

Reply:
Added a check for snapshot_info_cache_index != 0 before calling GetSnapshotInfoFromCache().  I also tagged the snapshotInfoCacheIndex parameter in GetInitiaitedSnapshotStrings and GetLatestConfigurationSnapshot methods with 'include_in_proto': False.  This parameter is only used to communicate between the above C functions and their corresponding GetSnapshotInfoFromCache function call.